optaplanner-core/src/main/java/org/optaplanner/core/impl/io/jaxb/JaxbIO.java

```diff
@@ -52,25 +79,125 @@ public JaxbIO(Class<T> rootClass, int indentation) {
         this.rootClass = rootClass;
         this.indentation = indentation;
         try {
-            JAXBContext jaxbContext = JAXBContext.newInstance(rootClass);
-            unmarshaller = jaxbContext.createUnmarshaller();
+            jaxbContext = JAXBContext.newInstance(rootClass);
             marshaller = jaxbContext.createMarshaller();
             marshaller.setProperty(Marshaller.JAXB_FORMATTED_OUTPUT, true);
             marshaller.setProperty(Marshaller.JAXB_ENCODING, StandardCharsets.UTF_8.toString());
         } catch (JAXBException jaxbException) {
-            String errMessage = String.format("Unable to create JAXB marshaller or unmarshaller for a root element class %s.",
+            String errMessage = String.format("Unable to create JAXB Marshaller for a root element class (%s).",
                     rootClass.getName());
-            throw new IllegalStateException(errMessage, jaxbException);
+            throw new OptaPlannerXmlSerializationException(errMessage, jaxbException);
         }
     }
 
     public T read(Reader reader) {
         Objects.requireNonNull(reader);
         try {
-            return (T) unmarshaller.unmarshal(reader);
+            @SuppressWarnings("unchecked")
+            final T object = (T) createUnmarshaller().unmarshal(reader);
+            return object;
         } catch (JAXBException jaxbException) {
-            String errMessage = String.format("Unable to read the %s from XML.", rootClass.getName());
-            throw new XmlUnmarshallingException(errMessage, jaxbException);
+            String errMessage = String.format(ERR_MSG_READ, rootClass.getName());
+            throw new OptaPlannerXmlSerializationException(errMessage, jaxbException);
+        }
+    }
+
+    public T readAndValidate(Reader reader, String schemaResource) {
+        Objects.requireNonNull(reader);
+        Objects.requireNonNull(schemaResource);
+        SchemaFactory schemaFactory = SchemaFactory.newInstance(XMLConstants.W3C_XML_SCHEMA_NS_URI);
+        Schema schema;
+        try {
+            schema = schemaFactory.newSchema(JaxbIO.class.getResource(schemaResource));
+        } catch (SAXException e) {
+            throw new IllegalArgumentException("Unable to read input schema resource (" + schemaResource + ")", e);
+        }
+
+        Unmarshaller unmarshaller = createUnmarshaller();
+        unmarshaller.setSchema(schema);
+        ValidationEventStringCollector validationEventHandler = new ValidationEventStringCollector();
+
+        try {
+            unmarshaller.setEventHandler(validationEventHandler);
+        } catch (JAXBException jaxbException) {
+            String errMessage = String.format("Unable to set validation event handler to the unmarshaller for "
+                    + "a root element class (%s).", rootClass.getName());
+            throw new OptaPlannerXmlSerializationException(errMessage, jaxbException);
+        }
```

Comment:
The `eventHandler` is relevant only for validation, so it cannot move inside `createUnmarshaller()`. Getting rid of the checked JAXBException is the only thing that `createUnmarshaller()` does. I can remove it, but it will add a couple of more exception translations.